postgis/lwgeom_hexgrid.c

```diff
@@ -0,0 +1,360 @@
+ * (at your option) any later version.
+ *
+ * PostGIS is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PostGIS.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **********************************************************************
+ *
+ * Copyright 2009 Mark Cave-Ayland <mark.cave-ayland@siriusit.co.uk>
+ * Copyright 2009-2017 Paul Ramsey <pramsey@cleverelephant.ca>
+ * Copyright 2018 Darafei Praliaskouski <me@komzpa.net>
+ *
+ **********************************************************************/
+
+#include "postgres.h"
+#include "fmgr.h"
+#include "funcapi.h"
+#include "access/htup_details.h"
+
+#include "../postgis_config.h"
+#include "lwgeom_pg.h"
+#include "liblwgeom.h"
+
+#include <math.h>
+#include <float.h>
+#include <string.h>
+#include <stdio.h>
+
+
+Datum ST_Hexagon(PG_FUNCTION_ARGS);
+Datum ST_Square(PG_FUNCTION_ARGS);
+Datum ST_ShapeGrid(PG_FUNCTION_ARGS);
+
+/* ********* ********* ********* ********* ********* ********* ********* ********* */
+
+/* XXX TO DO: what about negative sizes? */
+
+/* Ratio of hexagon edge length to height = 2*cos(pi/6) */
+#define HXR 1.7320508075688774
+
+static const double hex_x[] = {0.0, 1.0, 1.5,     1.0, 0.0, -0.5,     0.0};
+static const double hex_y[] = {0.0, 0.0, 0.5*HXR, HXR, HXR,  0.5*HXR, 0.0};
+
+static LWGEOM *
+hexagon(double origin_x, double origin_y, double edge, int cell_i, int cell_j, uint32_t srid)
+{
+	double height = edge * HXR;
+	POINT4D pt;
+	POINTARRAY **ppa = lwalloc(sizeof(POINTARRAY*));
+	POINTARRAY *pa = ptarray_construct(0, 0, 7);
+	uint32_t i;
+
+	double offset_x = origin_x + (1.5 * edge * cell_i);
+	double offset_y = origin_y + (height * cell_j) + (0.5 * height * (cell_i % 2));
+	pt.z = pt.m = 0.0;
+
+	for (i = 0; i < 7; ++i)
+	{
+		pt.x = edge * hex_x[i] + offset_x;
+		pt.y = edge * hex_y[i] + offset_y;
+		ptarray_set_point4d(pa, i, &pt);
+	}
+
+	ppa[0] = pa;
+	return lwpoly_as_lwgeom(lwpoly_construct(srid, NULL, 1 /* nrings */, ppa));
+}
+
+static int
+hexagon_grid_size(double edge, const GBOX *gbox, int *cell_width, int *cell_height)
+{
+	double hex_height = edge * HXR;
+	double bounds_width  = gbox->xmax - gbox->xmin;
+	double bounds_height = gbox->ymax - gbox->ymin;
+	int width  = (int)floor(bounds_width/(1.5*edge)) + 1;
+	int height = (int)floor(bounds_height/hex_height) + 1;
+	if (cell_width) *cell_width = width;
+	if (cell_height) *cell_height = height;
+	return width * height;
+}
+
+/* ********* ********* ********* ********* ********* ********* ********* ********* */
+
+static const double sqr_x[] = {0.0, 1.0, 1.0, 0.0, 0.0};
+static const double sqr_y[] = {0.0, 0.0, 1.0, 1.0, 0.0};
+
+static LWGEOM *
+square(double origin_x, double origin_y, double E, int cell_i, int cell_j, uint32_t srid)
```

Comment:
I have some issue with the parameter naming:
- origin_x and origin_y: They require looking at the code to know if the origin is used as top-left, bottom-right or whatever.
- `E`: interesting name for the size, I have no idea where it might come from.

Reply:
E = edge, in my hand-written notes on the math of hexagon generating, so. 
the problem isn't so much the word "origin" as the "are my cell coordinates cartesian or image coordinates" which is really at the root of it. right now they are cartesian, so "origin" means "cartesian origin", however... the comment about wanting to generate equivalent Google grids raises the idea of having them be image coordinates instead. yuck.